김나은/4월 5주차/BOJ_G2_17825_주사위윷놀이.java

```diff
@@ -0,0 +1,147 @@
+package algo;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+import java.util.StringTokenizer;
+
+public class BOJ_G2_17825_주사위윷놀이 {
+
+	static int[] dice, order; // 주사위, 말 4개의 이동순서
+	static Node[] horse;
+	static Node start;
+	static int ans;
+
+	static class Node {
+		int score;
+		boolean isEmpty, isFinish;
+		Node next, fastPath; // 파란색 칸일 때 2가지 방향 존재
+
+		public Node(int score) {
+			this.score = score;
+			this.isEmpty = true;
+		}
+
+		public Node addNext(int score) {
+			Node nextNode = new Node(score);
+			this.next = nextNode;
+			return nextNode;
+		}
+
+		public static Node getNode(Node start, int target) {
+			Node temp = start;
+			while (true) {
+				if (temp == null) return null;
+				if (temp.score == target) return temp;
+				temp = temp.next;
+			}
+		}
+	}
+
+	public static void main(String[] args) throws IOException {
+
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		StringTokenizer st = new StringTokenizer(br.readLine());
+
+		dice = new int[10];
+		for (int i = 0; i < 10; i++) {
+			dice[i] = Integer.parseInt(st.nextToken());
+		}
+
+		init();
+
+		order = new int[10];
+		horse = new Node[4];
+		perm(0);
+		System.out.println(ans);
+
+	}
+
+	private static void perm(int cnt) {
+		if (cnt == 10) {
+			ans = Math.max(ans, gameStart());
+			return;
+		}
+
+		for (int i = 0; i < 4; i++) {
+			order[cnt] = i;
+			perm(cnt + 1);
+		}
+	}
+
+	private static int gameStart() {
+		Arrays.fill(horse, start); // 시작 칸에 말 4개 배치
+
+		int score = 0;
+		for (int i = 0; i < 10; i++) {
+			Node cur = horse[order[i]]; // 순열에 할당된 순서대로 말 이동
+			cur.isEmpty = true;  // 현재 있는 칸을 비워줌
+
+			for (int j = 0; j < dice[i]; j++) { // 주사위에 나온 수만큼 이동
+				if (j == 0 && cur.fastPath != null) { // 이동을 시작하려는 처음 위치가 파란색 칸이면
+					cur = cur.fastPath; // 지름길로 이동
+				} else {
+					cur = cur.next;
+				}
+			}
```

Comment:
`j==0` 일 때만 작동하는 조건이라 반복문 밖으로 빼도 좋을 것 같아요
```suggestion
                      int start = 0;
                      if(cur.fastPath != null)  {
                              cur = cur.fastPath;
                              start++;
                      }
			for (int j = start; j < dice[i]; j++) { // 주사위에 나온 수만큼 이동
				cur = cur.next;
			}
```